### PR TITLE
Upgrade to aws go sdk v2 for cloudwatch logging driver

### PR DESCRIPTION
**Pull request by jchorl**
**- What I did**
Go SDK v1 is pretty outdated. In particular, it struggles with things like error handling and rate limits. I upgraded the awslogs logging driver to use the v2 SDK.

**- How I did it**
Bumped the imports, made a bunch of code mods, verified it works

**- How to verify it**
I verified in 2 ways.

Firstly, tests. Note that credential tests are very hard to verify now. It seems SDKv2 doesn't have a good way to tell your client to _only_ get creds. Also though, we mainly piggyback on aws libs to get creds, so I'm not sure we need to test this.

**I'm not confident this won't break someone. But I don't see a great way to test. I'm hoping someone can find a good way to validate that we haven't broken functionality.**

I also _changed_ the behaviour of the user agent tagging. I'm not sure this matters. We'll probably ruin some product-manager's day at AWS. But now we're using their blessed framework to add UA, so it might be okay?

Other than that, tests all pass with minor modifications (and no changes in behaviour to my knowledge).

I then went ahead and compiled dockerd and threw this on an aws box. Can confirm, for basic logging, this works great:
```
[ec2-user@ip-10-0-103-225 ~]$ docker run -it --rm --log-driver awslogs --log-opt awslogs-group=joshtest1 ubuntu:22.04 echo josh-what-123
```

I added some logging by dockerd in testing:
```
Sep 03 01:13:23 ip-10-0-103-225.us-west-2.compute.internal dockerd[4862]: time="2022-09-03T01:13:23.859298821Z" level=info msg=JOSH-CREATING
```

So clearly it called into my build. And then we see the logs in real AWS:
Screenshot 2022-09-03 at 01-14-11 CloudWatch Management Console

So I have some confidence that this change doesn't catastrophically break things.

**- Description for the changelog**
Upgrade awslogs loggging aws sdk version to v2

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thank you for contributing! It appears your commit message is missing a DCO sign-off,
causing the DCO check to fail.

We require all commit messages to have a `Signed-off-by` line with your name
and e-mail (see ["Sign your work"](https://github.com/moby/moby/blob/v20.10.11/CONTRIBUTING.md#sign-your-work)
in the CONTRIBUTING.md in this repository), which looks something like:

```
Signed-off-by: YourFirsName YourLastName <yourname@example.org>
```

There is no need to open a new pull request, but to fix this (and make CI pass),
you need to _amend_ the commit(s) in this pull request, and "force push" the amended
commit.

Unfortunately, it's not possible to do so through GitHub's web UI, so this needs
to be done through the git commandline.

You can find some instructions in the output of the DCO check (which can be found
in the "checks" tab on this pull request), as well as in the [Moby contributing guide](https://github.com/moby/moby/blob/v20.10.11/docs/contributing/set-up-git.md).

Steps to do so "roughly" come down to:

1. Set your name and e-mail in git's configuration:

    ```bash
    git config --global user.name "YourFirstName YourLastName"
    git config --global user.email "yourname@example.org"
    ```

    (Make sure to use your _real_ name (**not your GitHub username/handle**) and e-mail)


2. Clone your fork locally
3. Check out the branch associated with this pull request
4. Sign-off and amend the existing commit(s)

    ```bash
    git commit --amend --no-edit --signoff
    ```

    If your pull request contains multiple commits, either squash the commits (if
    needed) or sign-off each individual commit.

5. _Force push_ your branch to GitHub (using the `--force` or [`--force-with-lease`](https://stackoverflow.com/questions/52823692/git-push-force-with-lease-vs-force) flags) to update the pull request.


Sorry for the hassle (I wish GitHub would make this a bit easier to do), and let me know if you need help or more detailed instructions!

**Comment by jchorl** [comment]
Should be fixed :crossed_fingers:

**Comment by thaJeztah** [comment]
Thanks! I see there's one issue (haven't checked if it's an actual issue or a false positive);

```
daemon/logger/awslogs/cloudwatchlogs.go:66:2: `userAgentHeader` is unused (deadcode)
    userAgentHeader = "User-Agent"
    ^
```

Probably the first two commits should be squashed as well, as the updated vendor code is not yet used until the second commit. If it makes reviewing easier, it's probably fine to keep that until we merge (as long as we don't forget 😅)

**Comment by jchorl** [comment]
Fixed the lint error, and if any more crop up I'll try to grab those. Sorry bout that!

> Probably the first two commits should be squashed as well

Interesting, they take the PR down from ~500 files to ~2. I figured would be much easier to review. I also assumed you all squash on-merge. Anyway! Let me know how I can help shepherd this through. Happy to squash if/when it is helpful.

**Comment by jchorl** [comment]
I believe the one failing test is probably a false positive, cloudflare tls handshake error seems entirely unrelated

**Comment by samuelkarp** [comment]
@kzys might be able to help here?

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (kzys) [inline daemon/logger/awslogs/cloudwatchlogs_test.go:614]: For gorountine-starting code, how about using context.WithCancel instead? So we could cancel the goroutine at the end.

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (kzys) [inline daemon/logger/awslogs/cloudwatchlogs_test.go:1655]: @samuelkarp do you know the origin of these tests? Not so sure we can remove them.

**Review by kzys — CHANGES_REQUESTED** [review]
Thanks for the work @jchorl! This is something AWS employees (including myself) should do before customers...

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs.go:144]: Done

Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs.go:144]: (also, agree)

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs.go:733]: I think this must have been a copy/paste error. The `awsErr` iface changed, and I probably copy/pasted the whole group. I don't see a reason I would have replaced `logStreamName` with `logCreateGroup`.

Anyway, I fixed.

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs.go:764]: Same as above. Thanks for the CR, sorry for the error.

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs_test.go:614]: Yep

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs_test.go:123]: I do hardcode that, no?

```
	expectedUserAgentString := fmt.Sprintf(
		"%s/%s os/%s lang/go/%s md/GOOS/%s md/GOARCH/%s api/cloudwatchlogs/1.15.17 Docker/%s",
		aws.SDKName, aws.SDKVersion, runtime.GOOS, strings.TrimPrefix(runtime.Version(), "go"), runtime.GOOS, runtime.GOARCH, dockerversion.Version,
	)
```

Or did you mean in the comment?

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs_test.go:1655]: Tbc, I think it would be the responsible thing to do to keep these tests. I couldn't find a great way to test creds with sdkv2. Very open to suggestions.

I'm terrified that we'll merge and some bespoke feature of aws sdk will break, with pretty widespread blast-radius and slow release cadence....

Comment (samuelkarp) [reply daemon/logger/awslogs/cloudwatchlogs_test.go:1655]: This test was to verify the functionality of `awslogs-credentials-endpoint`, which is a non-standard credential provider added to the awslogs driver that ECS and Fargate both depend on.

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (samuelkarp) [reply daemon/logger/awslogs/cloudwatchlogs.go:381]: I'm not really remembering; it's possible that this (custom) provider did not have a default timeout.

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (kzys) [reply daemon/logger/awslogs/cloudwatchlogs_test.go:123]: Oh sorry. My comment had to be;

> *Can't* we hardcode `1.15.17` here?

Moreover, is 1.5.17 the version of SDK? Should we keep it somewhat backward-compatible? (cc @samuelkarp).

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (sparrc) [reply daemon/logger/awslogs/cloudwatchlogs_test.go:1655]: We do use that functionality to support the [ECS task execution role](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_execution_IAM_role.html), I see above in the unit tests that there is a somewhat similar-looking unit test (at [line ~208](https://github.com/moby/moby/pull/44087/files#diff-a15897c1d4ace995483d1ed2e6414291d7181a9821b777b72f743189311a3fc1R208)). 

That test is also using a test server and changing the endpoint. I believe the same concept should still apply here. Essentially when we have `awslogs-credentials-endpoint` we are expecting that the awslogs client will grab credentials from this endpoint, using the "http" credentials endpoint provider. This was using the ["endpoint creds" credential provider from the v1 sdk,](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/endpointcreds/) you can see where this was option was parsed and switched to endpointcreds [here in code](https://github.com/moby/moby/blob/a189651b4c772881ba87eec181312cd8bc20ec2e/daemon/logger/awslogs/cloudwatchlogs.go#L384-L395).

Anyways, maybe that was TMI, but from what I understand in your code you have ported to the v2 version of endpoint creds, so AFAICT this unit test should still be working? Was there something in particular you couldn't get working?

Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs_test.go]: It's a bit tough, because we can no longer reach into the client and grab the creds, like `creds, err := client.Config.Credentials.Get()`.

We could of course define a function like `getCredentialsProvider(info.Config[credentialsEndpointKey])` and write a test on `getCredentialsProvider`, but then there's no guarantee it actually gets applied to the client.

I rigged up a messy test that I think does test for the functionality we want to test for. Let me know what you think, and if you feel there's a better way to test this.

**Review by kzys — APPROVED** [review]
Looks good to me. Thank you again!

@samuelkarp @thaJeztah Would you like to squash the commits?

**Comment by thaJeztah** [comment]
Did a quick squash and rebase 👍

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (thaJeztah) [inline daemon/logger/awslogs/cloudwatchlogs.go:388]: Same for this one

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (thaJeztah) [inline daemon/logger/awslogs/cloudwatchlogs.go:391]: Same as the other error. Probably AWS and SDK should be capitalised though 🤔 

```go
		logrus.WithError(err).Error("Could not initialize AWS SDK config")
		return nil, errors.Wrap(err, "could not initialize AWS SDK config")
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Some minor nits about wrapping, errors, and fixing an (already existing) nit for returning a concrete type vs interface.

I have the changes made locally, so will push a commit for others to review 👍

**Comment by thaJeztah** [comment]
Left some minor suggestions, and pushed them as a commit; will do a squash if others agree with those changes 👍

**Comment by jchorl** [comment]
all lgtm! thanks for helping shepherd this through

**Comment by jchorl** [comment]
@thaJeztah @samuelkarp any chance we can push this through?

**Review by thaJeztah — APPROVED** [review]
I squashed the commits (left the patch I added in the comment above)


LGTM

**Comment by jchorl** [comment]
@thaJeztah there are now a bunch of merge conflicts. What's the path to getting this merged? I can resolve the conflicts but if I'm gonna do that work, I'd rather we get it into master soon after to avoid further conflicts.

**Review by corhere — CHANGES_REQUESTED** [review]
I've resolved the merge conflicts as a demonstration of good faith that this PR will be reviewed and merged in a timely manner.

Plumbing through and handling contexts correctly is a really hairy topic when there are long-lived background operations which outlive requests (read: function calls returning) and backoff-retries. I have spotted a few instances of subtle incorrectness in context usage, but it's not really worth the trouble to make it correct while the daemon/logger API is context-unaware.

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs.go:484]: Nice, tested with https://gist.github.com/jchorl/1f2f74b71311e0d91a69ed643dba0405 and it works

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs_test.go:137]: Mostly agreed. I assume a different team manages the sdk than cloudwatch metrics/UA behaviour. Even if upgrading the sdk, the cloudwatch team may appreciate fudging the new sdk to maintain uniformity with the old-style values. But yeah, if that's the case, they just shouldn't break/change functionality between sdk versions.

File: daemon/logger/awslogs/cloudwatchlogs_test.go
Comment (jchorl) [reply daemon/logger/awslogs/cloudwatchlogs_test.go:212]: Saw your suggestion after I implemented `atomic.Value`. Anyway, that'll make it easier to go to `atomic.Bool`.

I was thinking about race conditions. We know the handler must return for the test control-flow to hit that last `assert`. So the `val.Store` must have been called before the `val.Load`, so I think we're good here.

I'd rather avoid channels because they're not that hard to deadlock/hang with.

**Comment by jchorl** [comment]
@corhere thank you for the detailed feedback. I believe I've addressed all of it (sorry if I missed any!).

I compiled this and ran it on an ec2 instance, and it worked. It's tough to simulate the error handling bits though - hopefully testing covers the major cases there.

**Review by corhere — APPROVED** [review]
LGTM, thanks!

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (neersighted) [inline daemon/logger/awslogs/cloudwatchlogs.go:496]: We should prefer `errors.As` here, with a `var` declaration for `apiErr`.

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (neersighted) [inline daemon/logger/awslogs/cloudwatchlogs.go:535]: Prefer `errors.As` (see line 496).

**Review by neersighted — CHANGES_REQUESTED** [review]

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (corhere) [reply daemon/logger/awslogs/cloudwatchlogs.go:496]: `errors.As` would be subtly incorrect as it could further unwrap the `apiErr`, potentially resulting in a different error being matched than the one being logged. While unlikely in this particular case, directly asserting the type of `apiErro` also better expresses the intent that no further inspection of the error chain is desired.

File: daemon/logger/awslogs/cloudwatchlogs.go
Comment (corhere) [reply daemon/logger/awslogs/cloudwatchlogs.go:706]: [This is an existing idiom](https://github.com/search?type=code&q=%2Fif%20%5Cw%2B%20%3A%3D%20%5C(%5C*%3F%5Ba-zA-Z0-9.%5D%2B%5C)%5C(nil%5C)%3B%20errors.As%2F) which localizes the type of error to the same line as where its case is handled. Or to put it another way, this is the `errors.As` analogue to `switch err.(type)`

**Review by neersighted — APPROVED** [review]

**Pull request merged**
